### PR TITLE
[DistillationTrainer refactor] Add the ctor generation-supporting state (add-only)

### PR DESCRIPTION
**Pull request by qgallouedec**
*Item 25 (Group C — generation-supporting state) — stacked on PR 24 (#6511). Part of #6449.*

Second half of the "state before consumer" step: the ctor now holds everything GRPO's generation stack (items 26–30) and chunked loss (31) will read, so those port verbatim. Add-only — nothing new is consumed yet.

- Store the scalar generation attrs (from the item-24 config fields): `self.top_k`, `self.min_p`, `self.repetition_penalty`, `self.chat_template_kwargs`, `self.pad_to_multiple_of`, `self.shuffle_dataset`.
- Add GRPO's ctor infra, copied verbatim: `self.model_kwarg_keys` (inspect `forward` before PEFT wrapping — feeds `_get_last_hidden_state` at 31), `self._tokenizer` (ternary form; the full VLM/`ProcessorMixin` branch lands with VLM at 45), `set_seed(args.seed, device_specific=True)`, `self._dist = DistributedBackend(...)`, `self.model.add_model_tags(self._tag_names)`.
- Minimal rename `_last_vllm_sync_step` → `_last_loaded_step` so `_generate_single_turn` (26) can reference it. The throwaway buffer sync gate (deleted with the buffer at 29) keeps its `% vllm_sync_frequency` modulo untouched — aligning it now would be polish on doomed code and would prematurely orphan the config field.
- Imports: `inspect`, `set_seed` (from `accelerate.utils`, matching GRPO), `DistributedBackend`. (`numpy` / `profiling_context` land with their first consumer at item 26.)

No test — add-only ctor state, GRPO-shaped (covered when wired via the generation train tests, like GRPO).

Verified: ruff clean; `test_distillation_trainer.py` (41) green.

---

> [!NOTE]
> **Low Risk**
> Add-only scaffolding with minimal runtime impact; the only user-visible tweak is optional chat-template kwargs on prompt tokenization, plus device-specific seeding for distributed generation.
> 
> **Overview**
> **Prepares `DistillationTrainer` for the upcoming GRPO generation stack** by wiring constructor-only state that later items will consume—no new generation or loss paths yet.
> 
> The ctor now **persists extra generation-related config** (`top_k`, `min_p`, `repetition_penalty`, `chat_template_kwargs`, `pad_to_multiple_of`, `shuffle_dataset`) alongside the existing sampling fields. It also adds **GRPO-shaped infrastructure**: `model_kwarg_keys` from `inspect.signature` on `forward` (before PEFT), `self._tokenizer` (processor vs tokenizer), `DistributedBackend`, `model.add_model_tags`, and **`set_seed(..., device_specific=True)`** to avoid duplicate completions across processes when `num_generations` is large.
> 
> **Small immediate behavior touches:** conversational `_tokenize_prompts` forwards **`chat_template_kwargs`** into `apply_chat_template`; vLLM weight sync tracking is renamed **`_last_vllm_sync_step` → `_last_loaded_step`** with the same modulo gate.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6ca34ad4b7cab25c3701c0efab3a7cbee1dce071. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_trainer.py [hunk 1]
```diff
@@ -430,6 +443,13 @@ def __init__(
         # self.model_accepts_loss_kwargs to False to enable scaling.
         self.model_accepts_loss_kwargs = False
 
+        set_seed(args.seed, device_specific=True)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py]: Fixed in 128cb5e5

File: trl/experimental/distillation/distillation_trainer.py [hunk 2]
```diff
@@ -321,6 +323,14 @@ def __init__(
         else:
             model_name_or_path = model.config._name_or_path if model is not None else None
 
+        # Some models (SmolVLM/Idefics3) don't support `logits_to_keep` argument and error out if we pass it
+        # Inspect the forward method before we wrap the model with PEFT
+        self.model_kwarg_keys = (
+            inspect.signature(model.forward).parameters.keys()
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_trainer.py:329]: Nit: This call may now raise an obscure error if `model=None`. 

Previously, `model=None` fell through the transformer's `Trainer` and raised a clear error: "Trainer requires either a model or a model_init argument".

Now, it raises "'NoneType' object has no attribute 'forward'". The `DistillationTrainer` constructor still allows for `model=None`, so this scenario may still be reached.

I'd say this is non-blocking and out of scope for this PR, but maybe we can remove support for `model=None` in the refactoring plan as we do in the GRPO trainer.

**Review by cmpatino — APPROVED** [review]

File: trl/experimental/distillation/distillation_trainer.py (hunk 2 above)
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:329]: agreed it's out of scope here. I've added "make model required (drop | None = None, matching GRPO)" to the plan at item 40 (ctor signature cleanup), and the server subclass's own `model=None` default follows at item 55. #6449 updated.
